libgadget/forcetree.c

```diff
@@ -1149,74 +1131,57 @@ void force_update_hmax(int * activeset, int size, ForceTree * tree)
 
         while(no >= 0)
         {
-            if(P[p_i].Hsml <= tree->Nodes[no].u.d.hmax) break;
-
-            tree->Nodes[no].u.d.hmax = P[p_i].Hsml;
-
-            if(tree->Nodes[no].f.TopLevel) /* we reached a top-level node */
-            {
-                if (!tree->Nodes[no].f.NodeIsDirty) {
-                    tree->Nodes[no].f.NodeIsDirty = 1;
-                    DirtyTopLevelNodes[NumDirtyTopLevelNodes].treenode = no;
-                    DirtyTopLevelNodes[NumDirtyTopLevelNodes].hmax = tree->Nodes[no].u.d.hmax;
-                    NumDirtyTopLevelNodes ++;
-                }
+            if(P[p_i].Hsml <= tree->Nodes[no].u.d.hmax)
                 break;
-            }
-
+            tree->Nodes[no].u.d.hmax = P[p_i].Hsml;
             no = tree->Nodes[no].father;
         }
     }
 
-    /* share the hmax-data of the dirty nodes accross CPUs */
+    double * TopLeafhmax = (double *) mymalloc("TopLeafMoments", ddecomp->NTopLeaves * sizeof(double));
```

Comment:
The old algorithm only exchanges top leaves that are 'dirty'. Your new algorithm seems to be exchanging all TopLeaves. 

Do we already have MPI operations per step on the order of NTopLeaves? (We probably do, since the tree is built per step)